fvm/evm/emulator/state/delta.go

```diff
@@ -0,0 +1,414 @@
+		codeHashes:          make(map[gethCommon.Address]gethCommon.Hash, 0),
+		dirtySlots:          make(map[types.SlotAddress]interface{}, 0),
+		states:              make(map[types.SlotAddress]gethCommon.Hash, 0),
+		transient:           make(map[types.SlotAddress]gethCommon.Hash, 0),
+		accessListAddresses: make(map[gethCommon.Address]interface{}, 0),
+		accessListSlots:     make(map[types.SlotAddress]interface{}, 0),
+		logs:                make([]*gethTypes.Log, 0),
+		preimages:           make(map[gethCommon.Hash][]byte, 0),
+
+		// for refund we just copy the data
+		refund: parent.GetRefund(),
+	}
+}
+
+// NewChildView constructs a new delta view having the current view as parent
+func (d *DeltaView) NewChildView() *DeltaView {
+	return NewDeltaView(d)
+}
+
+// Exist returns true if address exists
+//
+// it also returns true for both newly created accounts or accounts that has been flagged for deletion
+func (d *DeltaView) Exist(addr gethCommon.Address) (bool, error) {
+	_, found := d.created[addr]
+	if found {
+		return true, nil
+	}
+	_, found = d.suicided[addr]
+	if found {
+		return true, nil
+	}
+	return d.parent.Exist(addr)
+}
+
+// CreateAccount creates a new account for the given address
+func (d *DeltaView) CreateAccount(addr gethCommon.Address) error {
+	d.created[addr] = struct{}{}
+	// flag the address as dirty
+	d.dirtyAddresses[addr] = struct{}{}
+	return nil
+}
+
+// IsCreated returns true if address has been created in this tx
+func (d *DeltaView) IsCreated(addr gethCommon.Address) bool {
+	_, found := d.created[addr]
+	if found {
+		return true
+	}
+	return d.parent.IsCreated(addr)
+}
+
+// HasSuicided returns true if address has been flagged for deletion
+func (d *DeltaView) HasSuicided(addr gethCommon.Address) bool {
+	_, found := d.suicided[addr]
+	if found {
+		return true
+	}
+	return d.parent.HasSuicided(addr)
+}
+
+// Suicide sets a flag to delete the account at the end of transaction
+func (d *DeltaView) Suicide(addr gethCommon.Address) (bool, error) {
+	// if it doesn't exist, return false
+	exists, err := d.Exist(addr)
+	if err != nil {
+		return false, err
+	}
+	if !exists {
+		return false, nil
+	}
+	// flag the account for deletion
+	d.suicided[addr] = struct{}{}
+
+	// flag the address as dirty
+	d.dirtyAddresses[addr] = struct{}{}
+	return true, nil
+}
+
+// GetBalance returns the balance of the given address
+func (d *DeltaView) GetBalance(addr gethCommon.Address) (*big.Int, error) {
+	if d.HasSuicided(addr) {
+		return big.NewInt(0), nil
+	}
+	val, found := d.balances[addr]
+	if found {
+		return val, nil
+	}
+	// if newly created and no balance is set yet
+	_, newlyCreated := d.created[addr]
+	if newlyCreated {
+		return big.NewInt(0), nil
+	}
+	return d.parent.GetBalance(addr)
+}
+
+// AddBalance adds the amount to the current balance of the given address
+func (d *DeltaView) AddBalance(addr gethCommon.Address, amount *big.Int) error {
+	// if amount is 0 skip
+	if amount.Sign() == 0 {
```

Comment:
why are we skipping if amount is 0, don't all accounts have first amount as 0 before the add balance is called?